modules/pam_oath/pam_oath.8
.\"-
.\" Copyright (c) 2012-2016 The University of Oslo
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"
.\" $Id$
.\"
.Dd January 11, 2016
.Dt PAM_OATH 8
.Os
.Sh NAME
.Nm pam_oath
.Nd OATH service module
.Sh SYNOPSIS
.Op Ar service
.Ar module-type
.Ar control-flag
.Cm Nm
.Op Ar arguments
.Sh DESCRIPTION
The
.Nm
service module implements counter-based and time-based one-time
passwords.
.Pp
The
.Nm
service module recognizes the following options:
.Bl -tag -width ".Cm echo_pass"
.It Cm nokey = Ar fail | fake | ignore
Specifies how the module should behave when no key is available for
the user: either fail immediately, prompt for a code but fail anyway,
or let authentication proceed by other means.
\" .It Cm nouser = Ar fail | fake | ignore
\" Specifies how the module should behave when the user does not exist.
\" See
\" .Bm nokey
\" above.
\" .It Cm badkey = Ar fail | fake | ignore
\" Specifies how the module should behave when the user exists and has a
\" key, but the key could not be loaded (e.g. due to a syntax error in
\" the keyfile).
\" See
\" .Bm nokey
\" above.
.El
.Pp
The
.Nm
service module uses
.Xr pam_get_authtok 3
to prompt the user, and will therefore also be affected by the
standard options
.Sh SEE ALSO
.Xr oathkey 1 ,
.Xr pam.conf 5 ,
.Xr pam 8
.Sh AUTHORS
The
.Nm
module and this manual page were developed by
.An Dag-Erling Sm\(/orgrav Aq des@des.no
for the University of Oslo.


HISTORY
OpenPAM ??????????						2014-??-??

 - BUGFIX: Under certain circumstances, specifying a non-existent
   module (or misspelling the name of a module) in a policy could
   result in a fail-open scenario.

 - FEATURE: Add a pam_oath module that implements RFC 4226 (HOTP) and
   RFC 6238 (TOTP).

 - BUGFIX: The is_upper() predicate only accepted the letter A as an
   upper-case character instead of the entire A-Z range.  As a result,
   service and module names containing upper-case letters other than A
   would be rejected.
============================================================================
OpenPAM Nummularia						2013-09-07

 - ENHANCE: Rewrite the dynamic loader to improve readability and
   reliability.  Modules can now be listed without the ".so" suffix in
   the policy file; OpenPAM will automatically add it, just like it
   will automatically add the version number if required.

 - ENHANCE: Allow openpam_straddch(3) to be called without a character
   so it can be used to preallocate a string.

 - ENHANCE: Improve portability by adding simple asprintf(3) and
   vasprintf(3) implementations for platforms that don't have them.

 - ENHANCE: Move the libpam sources into a separate subdirectory.

 - ENHANCE: Substantial documentation improvements.

 - BUGFIX: When openpam_readword(3) encountered an opening quote, it
   would set the first byte in the buffer to '\0', discarding all
   existing text and, unless the buffer was empty to begin with, all
   subsequent text as well.  This went unnoticed because none of the
   unit tests for quoted strings had any text preceding the opening
   quote.

 - BUGFIX: make --with-modules-dir work the way it was meant to work
   (but never did).
============================================================================
OpenPAM Micrampelis						2012-05-26

 - FEATURE: Add an openpam_readword(3) function which reads the next
   word from an input stream, applying shell quoting and escaping
   rules.  Add numerous unit tests for openpam_readword(3).

 - FEATURE: Add an openpam_readlinev(3) function which uses the
   openpam_readword(3) function to read words from an input stream one
   at a time until it reaches an unquoted, unescaped newline, and
   returns an array of those words.  Add several unit tests for
   openpam_readlinev(3).

 - FEATURE: Add a PAM_HOST item which pam_start(3) initializes to the
   machine's hostname.  This was implemented in Lycopsida but
   inadvertantly left out of the release notes.

 - FEATURE: In pam_get_authtok(3), if neither the application nor the
   module have specified a prompt and PAM_HOST and PAM_RHOST are both
   defined but not equal, use a different default prompt that includes
   PAM_USER and PAM_HOST.

 - ENHANCE: Rewrite the policy parser to used openpam_readlinev(),
   which greatly simplifies the code.

 - ENHANCE: The previous implementation of the policy parser relied on
   the openpam_readline(3) function, which (by design) munges
   whitespace and understands neither quotes nor backslash escapes.
   As a result of the aforementioned rewrite, whitespace, quotes and
   backslash escapes in policy files are now handled in a consistent
   and predictable manner.

 - ENHANCE: On platforms that have it, use fdlopen(3) to load modules.
   This closes the race between the ownership / permission check and
   the dlopen(3) call.

 - ENHANCE: Reduce the amount of pointless error messages generated
   while searching for a module.

 - ENHANCE: Numerous documentation improvements, both in content and
   formatting.

 - BUGFIX: A patch incorporated in Lycopsida inadvertantly changed
   OpenPAM's behavior when several policies exist for the same
   service, from ignoring all but the first to concatenating them all.
   Revert to the original behavior.

 - BUGFIX: Plug a memory leak in the policy parser.
============================================================================
OpenPAM Lycopsida						2011-12-18

 - ENHANCE: removed static build autodetection, which didn't work
   anyway.  Use an explicit, user-specified preprocessor variable
   instead.

 - ENHANCE: cleaned up the documentation a bit.

 - ENHANCE: added openpam_subst(3), allowing certain PAM items to be
   embedded in strings such as prompts.  Apply it to the prompts used
   by pam_get_user(3) and pam_get_authtok(3).

 - ENHANCE: added support for the user_prompt, authtok_prompt and
   oldauthtok_prompt module options, which override the prompts passed
   by the module to pam_set_user(3) and pam_get_authtok(3).

 - ENHANCE: rewrote the policy parser to support quoted option values.

 - ENHANCE: added pamtest(1), a tool for testing modules and policies.

 - ENHANCE: added code to check the ownership and permissions of a
   module before loading it.

 - ENHANCE: added / improved input validation in many cases, including
   the policy file and some function arguments.
============================================================================
OpenPAM Hydrangea						2007-12-21

 - ENHANCE: when compiling with GCC, mark up API functions with GCC
   attributes where appropriate.

 - BUGFIX: fixed numerous warnings uncovered by GCC 4.

 - ENHANCE: building the documentation is now optional.

 - ENHANCE: corrected a number of mistakes and style issues in the
   build system.

 - ENHANCE: API function arguments are now const where appropriate, to
   match corresponding changes in the Solaris PAM and Linux-PAM APIs.

 - ENHANCE: corrected a number of C namespace violations.

 - ENHANCE: the module cache has been removed, allowing long-lived
   applications to pick up module changes.  This also allows multiple
   threads to use PAM simultaneously (as long as they use separate PAM
   contexts), since the module cache was the only part of OpenPAM that
   was not thread-safe.
============================================================================
OpenPAM Figwort							2005-06-16

 - BUGFIX: Correct several small signedness and initialization bugs

trunk/modules/pam_oath/pam_oath.c
/*-
 * Copyright (c) 2012-2013 The University of Oslo
 * Copyright (c) 2016 Dag-Erling Smørgrav
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id$
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <fcntl.h>
#include <limits.h>
#include <pwd.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define PAM_SM_AUTH
#define PAM_SM_ACCOUNT

#include <security/pam_modules.h>
#include <security/pam_appl.h>
#include <security/oath.h>

#define PAM_OATH_PROMPT "Verification code: "
#define PAM_OATH_HOTP_WINDOW 3
#define PAM_OATH_TOTP_WINDOW 3

enum pam_oath_nokey { nokey_error = -1, nokey_fail, nokey_fake, nokey_ignore };

static const char *pam_oath_default_keyfile = "/var/oath/%u.otpauth";

/*
 * Parse the nokey or badkey option, which indicates how we should act if
 * the user has no keyfile or the keyfile is invalid.
 */
static enum pam_oath_nokey
pam_oath_nokey_option(pam_handle_t *pamh, const char *option)
{
	const char *value;

	if ((value = openpam_get_option(pamh, option)) == NULL)
		return (nokey_fail);
	else if (strcmp(value, "fail") == 0)
		return (nokey_fail);
	else if (strcmp(value, "fake") == 0)
		return (nokey_fake);
	else if (strcmp(value, "ignore") == 0)
		return (nokey_ignore);
	openpam_log(PAM_LOG_ERROR, "the value of the %s option "
	    "must be either 'fail', 'fake' or 'ignore'", option);
	return (nokey_error);
}

/*
 * Parse a numeric option.  Returns -1 if the option is not set or its
 * value is not an integer in the range [0, INT_MAX].
 */
static int
pam_oath_int_option(pam_handle_t *pamh, const char *option)
{
	const char *value;
	char *end;
	long num;

	if ((value = openpam_get_option(pamh, option)) == NULL)
		return (-1);
	num = strtol(value, &end, 10);
	if (*value == '\0' || *end != '\0' || num < 0 || num > INT_MAX) {
		openpam_log(PAM_LOG_ERROR, "the value of the %s option "
		    "is invalid.", option);
		return (-1);
	}
	return (num);
}

/*
 * Determine the location of the user's keyfile.
 */
static char *
pam_oath_keyfile(pam_handle_t *pamh)
{
	const char *keyfile;
	char *path;
	size_t size;

	if ((keyfile = openpam_get_option(pamh, "keyfile")) == NULL)
		keyfile = pam_oath_default_keyfile;
	size = 0;
	if (openpam_subst(pamh, NULL, &size, keyfile) != PAM_TRY_AGAIN)
		return (NULL);
	if ((path = malloc(size)) == NULL)
		return (NULL);
	if (openpam_subst(pamh, path, &size, keyfile) != PAM_SUCCESS) {
		free(path);
		return (NULL);
	}
	return (path);
}

/*
 * Load the user's key.
 */
static struct oath_key *
pam_oath_load_key(const char *keyfile)
{

	/* XXX should check ownership and permissions */
	return (oath_key_from_file(keyfile));
}

/*
 * Save the user's key.
 * XXX should be a liboath API function.
 */
static int
pam_oath_save_key(const struct oath_key *key, const char *keyfile)
{
	char *keyuri;
	int fd, len, pam_err;

	keyuri = NULL;
	len = 0;
	fd = -1;
	pam_err = PAM_SYSTEM_ERR;
	if ((keyuri = oath_key_to_uri(key)) == NULL)
		goto done;
	len = strlen(keyuri);
	if ((fd = open(keyfile, O_WRONLY|O_CREAT|O_TRUNC, 0600)) < 0 ||
	    write(fd, keyuri, len) != len || write(fd, "\n", 1) != 1) {
		openpam_log(PAM_LOG_ERROR, "%s: %m", keyfile);
		goto done;
	}
	pam_err = PAM_SUCCESS;
done:
	if (fd >= 0)
		close(fd);
	if (keyuri != NULL) {
		memset(keyuri, 0, len);
		free(keyuri);
	}
	return (pam_err);
}

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags,
	int argc, const char *argv[])
{
	enum pam_oath_nokey nokey, badkey;
	struct passwd *pwd;
	const char *user, *password;
	char *end, *keyfile;
	struct oath_key *key;
	unsigned long response;
	int pam_err, ret, window;

	/* unused */
	(void)flags;
	(void)argc;
	(void)argv;

	keyfile = NULL;
	key = NULL;

	openpam_log(PAM_LOG_VERBOSE, "attempting OATH authentication");

	/* check how to behave if the user does not have a valid key */
	if ((nokey = pam_oath_nokey_option(pamh, "nokey")) == nokey_error ||
	    (badkey = pam_oath_nokey_option(pamh, "badkey")) == nokey_error) {
		pam_err = PAM_SERVICE_ERR;
		goto done;
	}

	/* identify user */
	if ((pam_err = pam_get_user(pamh, &user, NULL)) != PAM_SUCCESS)
		goto done;
	if ((pwd = getpwnam(user)) == NULL) {
		pam_err = PAM_USER_UNKNOWN;
		goto done;
	}

	openpam_log(PAM_LOG_VERBOSE, "authenticating user %s", user);

	/* load key */
	if ((keyfile = pam_oath_keyfile(pamh)) == NULL) {
		pam_err = PAM_SYSTEM_ERR;
		goto done;
	}
	openpam_log(PAM_LOG_VERBOSE, "attempting to load %s for %s", keyfile, user);
	key = pam_oath_load_key(keyfile);

	/*
	 * The user doesn't have a key, should we fake it?
	 *
	 * XXX implement badkey - currently, oath_key_from_file() doesn't
	 * provide enough information for us to tell the difference
	 * between a bad key and no key at all.
	 *
	 * XXX move this into pam_oath_load_key()
	 */
	if (key == NULL) {
		openpam_log(PAM_LOG_VERBOSE, "no key found for %s", user);
		switch (nokey) {
		case nokey_fail:
			pam_err = PAM_AUTHINFO_UNAVAIL;
			goto done;
		case nokey_fake:
			key = oath_key_dummy(om_hotp, oh_sha1, 6);
			break;
		case nokey_ignore:
			pam_err = PAM_IGNORE;
			goto done;
		default:
			/* can't happen */
			pam_err = PAM_SERVICE_ERR;
			goto done;
		}
	}

	/* get user's response */
	pam_err = pam_get_authtok(pamh, PAM_AUTHTOK,
	    &password, PAM_OATH_PROMPT);
	if (pam_err != PAM_SUCCESS) {
		openpam_log(PAM_LOG_VERBOSE, "conversation failure");
		goto done;
	}

	/* convert to number */
	response = strtoul(password, &end, 10);
	if (end == password || *end != '\0')
		response = ULONG_MAX;

	/* verify response */
	if (key->mode == om_hotp) {
		if ((window = pam_oath_int_option(pamh, "hotp_window")) < 0 &&
		    (window = pam_oath_int_option(pamh, "window")) < 0)
			window = PAM_OATH_HOTP_WINDOW;
		ret = oath_hotp_match(key, response, window);
	} else {
		if ((window = pam_oath_int_option(pamh, "totp_window")) < 0 &&
		    (window = pam_oath_int_option(pamh, "window")) < 0)
			window = PAM_OATH_TOTP_WINDOW;
		ret = oath_totp_match(key, response, window);
	}
	openpam_log(PAM_LOG_VERBOSE, "verification code %s",
	    ret > 0 ? "matched" : "did not match");
	if (ret <= 0) {
		pam_err = ret < 0 ? PAM_SERVICE_ERR : PAM_AUTH_ERR;
		goto done;
	}

	/* write back the key (update counter for HOTP, lastused for TOTP) */
	if (pam_oath_save_key(key, keyfile) != 0) {
		pam_err = PAM_SERVICE_ERR;
		goto done;
	}

	openpam_log(PAM_LOG_VERBOSE, "OATH authentication succeeded");
	pam_err = PAM_SUCCESS;
done:
	oath_key_free(key);
	free(keyfile);
	return (pam_err);
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh, int flags,
	int argc, const char *argv[])
{

	/* unused */
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;
	return (PAM_SUCCESS);
}

PAM_EXTERN int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags,
	int argc, const char *argv[])
{

	/* unused */
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;
	return (PAM_SUCCESS);
}

PAM_MODULE_ENTRY("pam_unix");
